apollo-router/src/plugins/connectors/handle_responses.rs

```diff
@@ -58,9 +59,11 @@ pub(crate) async fn handle_responses(
 
         if parts.status.is_success() {
             let Ok(json_data) = serde_json::from_slice::<Value>(body) else {
-                if let Some(ref mut debug) = debug {
-                    debug.push_invalid_response(&parts, body);
-                }
+                context.extensions().with_lock(|mut lock| {
+                    if let Some(ref mut debug) = lock.get_mut::<ConnectorContext>() {
+                        debug.push_invalid_response(&parts, body);
+                    }
+                });
```

Comment:
I worry about lock contention—especially since it could slow down _all_ requests even if debug is disabled. Can we get router team eyes on this for a suggestion? Maybe we could make extensions `RwLock`, or maybe we should be storing an `Option<Mutex<ConnectorContext>>` and getting a reference to that (so we can still check for debug disabled).

Reply:
Sure, I can reach out for suggestions. For your second proposal, storing `Option<Mutex<ConnectorContext>>` would still require locking the `ExtensionsGuard` to get that `Option`, so there's still potential contention there, or am I misunderstanding?

Reply:
Actually, I think I see - you mean we could get that `Option` once and then if `Some`, lock the second `Mutex` inside the `Option`. So we'd only lock the extensions guard once instead of several times per request.

Reply:
And because the Option is on the outside of the second mutex, we wouldn't need to lock it at all if debugging is disabled

Reply:
I like this option the best and coded it up.